### PR TITLE
#229. Implement sign up screen.

### PR DESCRIPTION
**Pull request by bogdan-evtushenko**

### CODE REVIEW

File: app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignUpActivity.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        tvSignIn.setOnClickListener { finish() }
+    }
+
+    private fun signUp() {
+        val email = ifEmail.editText.text.trim().toString()
+        val password = ifPassword.editText.text.toString()
+        val confirmedPassword = ifConfirmPassword.editText.text.toString()
+        when {
+            password != confirmedPassword -> store.dispatch(AuthRequests.SignUp.Failure(getString(R.string.passwords_do_not_match).toMessage()))
+            checkbox.isChecked.not() -> store.dispatch(AuthRequests.SignUp.Failure(getString(R.string.agree_to_the_privacy_policy).toMessage()))
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignUpActivity.kt:70]: `.not()` is considered a bad practice (even though I like it too), so it's more common to use `!`.

File: app/src/main/java/com/bogdan/codeforceswatcher/util/Strings.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package com.bogdan.codeforceswatcher.util
+
+import android.text.Spannable
+import android.text.SpannableString
+import android.text.style.ForegroundColorSpan
+import androidx.core.content.ContextCompat
+import com.bogdan.codeforceswatcher.CwApp
+import com.bogdan.codeforceswatcher.R
+
+fun SpannableString.colorSubstring(l: Int, r: Int, color: Int) {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/util/Strings.kt:10]: Shouldn't it be `private`?

File: app/src/main/java/com/bogdan/codeforceswatcher/util/Strings.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    val foregroundColorSpan = ForegroundColorSpan(ContextCompat.getColor(CwApp.app, color))
+    this.setSpan(foregroundColorSpan, l, r, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+}
+
+fun String.colorLinkTagPart(): SpannableString {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/util/Strings.kt:15]: Name IMHO can be simplified to `linked()`.

File: common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AutghReducer.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package io.xorum.codeforceswatcher.features.auth
+
+import io.xorum.codeforceswatcher.redux.states.AppState
+import tw.geothings.rekotlin.Action
+
+fun authReducer(action: Action, state: AppState): AuthState {
+    var newState = state.auth
+
+    when (action) {
+
```
Comment (yevhenii-kanivets) [inline common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AutghReducer.kt:10]: Nitpick: don't like this empty line.

File: common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AutghReducer.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        }
+        is AuthRequests.SignIn.Failure -> {
+            newState = newState.copy(
+                    signInStatus = AuthState.Status.IDLE
+            )
+        }
+
+        is AuthRequests.SignUp.Success -> {
+            newState = newState.copy(
+                    signUpStatus = AuthState.Status.DONE,
```
Comment (yevhenii-kanivets) [inline common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AutghReducer.kt:29]: Nice 👍

**Review by yevhenii-kanivets — CHANGES_REQUESTED** [review]
Almost perfect 👍 Just a few minor points:
- "Sign Up" button should have disabled state until "privacy policy" isn't agreed
- Links to policy and terms seems to be not clickable

File: app/src/main/java/com/bogdan/codeforceswatcher/util/Strings.kt (hunk 2 above)
Comment (bogdan-evtushenko) [reply app/src/main/java/com/bogdan/codeforceswatcher/util/Strings.kt:10]: Nope, it's used in a few more cases.

**Review by yevhenii-kanivets — APPROVED** [review]
Thanks!

**Pull request merged**